salt/_modules/caasp_cri.py

```diff
@@ -0,0 +1,263 @@
+from __future__ import absolute_import
+
+import json
+import logging
+import os.path
+import time
+from salt.exceptions import CommandExecutionError
+
+try:
+    from salt.exceptions import InvalidConfigError
+except ImportError:
+    from salt.exceptions import SaltException
+
+    class InvalidConfigError(SaltException):
+        '''
+        Not yet defined by this version of salt
+        '''
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    return "caasp_cri"
+
+
+def cri_name():
+    '''
+    Calculate the CRI name by looking at the pillar set by the user.
+
+    Forces the 'docker' CRI to be used on the nodes that have certain roles,
+    this is needed because salt pillars exposed by Velum have precedence
+    over everything.
+    '''
+    roles_requiring_docker = ('admin', 'ca')
+
+    node_roles = __salt__['grains.get']('roles', [])
+    cri = __salt__['pillar.get']('cri:name', 'docker').lower()
+
+    for role in node_roles:
+        if role in roles_requiring_docker:
+            return 'docker'
+
+    return cri
+
+
+def get_container_id(name, namespace):
+    '''
+    Return the ID of the running container named ``name`` running inside of
+    the specified ``namespace``.
+
+    Returns ``None`` if no running container was found matching the criteria.
+
+    name
+        Name of the container. This is checked against the ``metadata.name``
+        field of a kubernetes pod.
+
+    namespace
+        Name of the namespace to search the container inside.
+
+    CLI example:
+
+    .. code-block:: bash
+
+        salt '*' caasp_cri.get_container_id name='haproxy' namespace='kube-system'
+    '''
+
+    __wait_CRI_socket()
+    cmd = "crictl --runtime-endpoint {socket} ps -o json".format(
+            socket=cri_runtime_endpoint())
+    result = __salt__['cmd.run_all'](cmd,
+                                     output_loglevel='trace',
+                                     python_shell=False)
+
+    if result['retcode'] != 0:
+        raise CommandExecutionError(
+                'Could not invoke crictl',
+                info={'errors': [result['stderr']]}
+            )
+
+    ps_data = json.loads(result['stdout'])
```

Comment:
The function returns the `ID` of the container of `None` when no running container is found. I decided to not handle the exception to let everything blow up. I just didn't want to swallow the exception into a log file.

Probably the best approach would be to handle the exception at a higher level (like the salt state actually using it) to return a better error message to the end user... Thoughts?

Reply:
I agree with you about sending an exception to the upper layers, but I think it should be some exception specific to CRI (ie, `CRIRuntimeException`) instead of a JSON decoding error that means nothing to a user of this module. Anyway, it is a small detail...